### DISCUSSION TITLE
Part and cost reduction

### DISCUSSION THREAD

**Discussion by bob-u** [tjko/fanpico#17] [Ideas]
Great project.  This is basically a fan proxy - very similar to what I've been thinking of making.  I'm glad I found this project and thank you for sharing.  Given the part shortages, what do you think about also using 74HC4051D for reading PWM input from the motherboard and substituting level converters with PC817 opto-couplers?

I also want to present the temperature sensor output to lm_sensors via USB.

**Comment by tjko** [comment]
I have had no issues getting any of the ICs, but I guess things are not 'stable' yet.... I went with these since at the time I was designing the initial schematic these were in plenty of supply and they conveniently had ESD protection built-in, etc...

I don't see problem adding support for boards to use multiplexer for (motherboard) PWM input. There is already similarly support for two different ways for boards to implement reading tachometer signals (directly or via multiplexer)...

Were you thinking of putting optocouplers before the 74HC4051 (one for each input)?

For lm_sensors, where you thinking making device appear as 'hwmon' device in USB bus? Or adding lm_sensor 'driver' that would simply issue SCPI commands to get the data?

**Comment by bob-u** [comment]
For interface I was thinking of:
USBTMC via  tinyusb for pipico: [tinyusb](https://github.com/hathach/tinyusb)
liquidctl for client: [liquidctl](https://github.com/liquidctl/liquidctl)
liquidtux for hwmon: [liquidtux](https://github.com/liquidctl/liquidtux)

EDIT: I did not realize that fanpico has a fully working SCPI interface already.  There's hwmon scpi module but in 2017 it was restricted to ARM builds only.  Still, it would probably be easier to change the complie flag then add new code into kernel (what some had been trying for years).

**Comment by Stoatwblr** [comment]
Very late to the party but wanting a similar end result...

As long as fanpico's SCPI is compliant to the API then it can be treated as a generic SCPI device

This basic script should do the trick once tweaked for the right access mode

```
python

import time
import os
import pyvisa

# 1. Connect to your SCPI test bench instrument
rm = pyvisa.ResourceManager('@py')
# Replace with your actual USB/Ethernet SCPI string
instrument = rm.open_resource('TCPIP0::192.168.1.150::5025::SOCKET') 

# 2. Setup a fake sysfs-style path for lm_sensors to grab
FAKE_SYSFS = "/tmp/scpi_hwmon"
os.makedirs(FAKE_SYSFS, exist_ok=True)

# Create the basic identity files lm_sensors expects
with open(f"{FAKE_SYSFS}/name", "w") as f:
    f.write("scpi_instrument\n")

while True:
    try:
        # 3. Request the data from your SCPI hardware
        # Example: Multimeter reading voltage or power supply reading current
        raw_voltage = instrument.query('MEASure:VOLTage:DC?') 
        
        # 4. Process and format for lm_sensors
        # lm_sensors expects voltages in millivolts (e.g., 5V = 5000)
        millivolts = int(float(raw_voltage) * 1000)
        
        # Write to the fake device node layout
        with open(f"{FAKE_SYSFS}/in0_input", "w") as f:
            f.write(f"{millivolts}\n")
            
        with open(f"{FAKE_SYSFS}/in0_label", "w") as f:
            f.write("Bench_Supply_V\n")
            
    except Exception as e:
        print(f"SCPI Error: {e}")
        
    time.sleep(1) # Poll every second
```
Symlink the output into sysfs hwmon and it should be picked up automatically (this will need to be repeated at each boot and before any daemons are started)

```
sudo ln -s /tmp/scpi_hwmon /sys/class/hwmon/hwmon99
```

Based on the kits, interfacing as a USB serial device is probably the way to fly

**Reply by tjko** [reply]
I seem to recall that no one (not even root) can write to _sysfs_?   So creating symlink  (_/sys/class/hwmon/hwmon99_) will fail(?)
